### PR TITLE
Backport DHT routing support to master-n3

### PR DESCRIPTION
**Pull request by Jim8y**
## Summary
- Backport signed node identity in the version handshake from #4420.
- Backport DHT routing table primitives and peer contact tracking from #4422.
- Correct closest-node selection to use full XOR-distance ordering before truncating results.
- Adapt the backport to the current master-n3 namespace and import style.

## Testing
- dotnet test tests/Neo.UnitTests/Neo.UnitTests.csproj --configuration Release --no-restore --filter "FullyQualifiedName~FindClosest_ConsidersAllBucketsWhenLowerBucketIsCloser" --verbosity minimal --disable-build-servers
- dotnet test tests/Neo.UnitTests/Neo.UnitTests.csproj --configuration Release --no-restore --filter "FullyQualifiedName~UT_RoutingTable|FullyQualifiedName~UT_UInt256" --verbosity minimal --disable-build-servers
- dotnet test tests/Neo.UnitTests/Neo.UnitTests.csproj --configuration Release --no-restore --filter "FullyQualifiedName~Network.P2P|FullyQualifiedName~UT_UInt256" --verbosity minimal --disable-build-servers
- dotnet test tests/Neo.UnitTests/Neo.UnitTests.csproj --configuration Release --no-restore --verbosity minimal --disable-build-servers
- dotnet format src/Neo/Neo.csproj --verify-no-changes --no-restore --verbosity minimal
- dotnet format tests/Neo.UnitTests/Neo.UnitTests.csproj --verify-no-changes --no-restore --verbosity minimal
- git diff --check

### CODE REVIEW

File: src/Neo/Network/P2P/Payloads/VersionPayload.cs [hunk 1]
```diff
@@ -101,7 +124,8 @@ void ISerializable.Deserialize(ref MemoryReader reader)
             Network = reader.ReadUInt32();
             Version = reader.ReadUInt32();
             Timestamp = reader.ReadUInt32();
-            Nonce = reader.ReadUInt32();
+            NodeKey = reader.ReadSerializable<ECPoint>();
+            NodeId = reader.ReadSerializable<UInt256>();
```
Comment (AnnaShaleva) [inline src/Neo/Network/P2P/Payloads/VersionPayload.cs:128]: It's an incompatible change, new nodes won't be able to handshake with the old nodes, it will affect the network connectivity on upgrade. At least we need some transition period and migration code.

**Comment by roman-khimov** [comment]
Why? N3 doesn't need it now and if we want to introduce it then some carefully planned transition period is needed.

**Comment by shargon** [comment]
Agree with @roman-khimov ,  planned transition period is needed

File: src/Neo/Network/P2P/Payloads/VersionPayload.cs [hunk 2]
```diff
@@ -101,7 +124,8 @@ void ISerializable.Deserialize(ref MemoryReader reader)
             Network = reader.ReadUInt32();
             Version = reader.ReadUInt32();
             Timestamp = reader.ReadUInt32();
-            Nonce = reader.ReadUInt32();
+            NodeKey = reader.ReadSerializable<ECPoint>();
```
Comment (cschuchardt88) [inline src/Neo/Network/P2P/Payloads/VersionPayload.cs:127]: **bug:** This is a hard wire break with no transition. `Nonce` (u32) is replaced by compressed `NodeKey` + `NodeId` + a var-size `Signature`, while `LocalNode.ProtocolVersion` stays `0`.

A current N3 node sending the old layout will fail `ECPoint`/`UInt256` deserialize (or produce garbage and fail `Verify`). A new node sending this layout cannot handshake with old nodes. There is no dual-decode path, no capability flag, and no version gate.

N3 mainnet cannot absorb this on upgrade. Keep the existing `Nonce` layout (or dual-decode by payload length / protocol version) until a planned cutover; do not ship signed identity as a silent `VersionPayload` resize.

File: src/Neo/Network/P2P/Payloads/VersionPayload.cs [hunk 3]
```diff
@@ -112,17 +136,34 @@ void ISerializable.Deserialize(ref MemoryReader reader)
...
             writer.Write(Timestamp);
-            writer.Write(Nonce);
+            writer.Write(NodeKey);
+            writer.Write(NodeId);
             writer.WriteVarString(UserAgent);
             writer.Write(Capabilities);
+            if (withSignature) writer.WriteVarBytes(Signature);
+        }
+
+        public bool Verify(ProtocolSettings protocol)
```
Comment (cschuchardt88) [inline src/Neo/Network/P2P/Payloads/VersionPayload.cs:160]: **bug:** `Verify` only checks `NodeId == NodeKey.GetNodeId(protocol)` and a self-signature of this payload. Nothing binds the signature to *this* TCP connection: no remote challenge/nonce, no listener endpoint, no timestamp window.

A captured `Version` message can be replayed on a new connection and will verify. `AllowNewConnection` only rejects the same `NodeId` on the same IP, so a replay from another address is accepted as that node (second `TaskManager.Register`, routing-table pollution).

If this identity is meant to be cryptographic, sign a handshake challenge (or at least reject a second live connection with the same `NodeId` regardless of IP, and enforce a timestamp skew).

File: src/Neo/Network/P2P/RemoteNode.cs [hunk 4]
```diff
@@ -140,6 +140,12 @@ protected override void OnAck()
             CheckMessageQueue();
         }
 
+        protected override void OnDisconnect(DisconnectReason reason)
```
Comment (cschuchardt88) [inline src/Neo/Network/P2P/RemoteNode.cs:143]: **bug:** `MarkFailure` only runs from `Connection.Disconnect()`. `Tcp.ConnectionClosed` stops the actor without calling `Disconnect`/`OnDisconnect` (`disconnected` stays false; `PostStop` only sends `Tcp.Close`).

A peer that drops the socket is left in `RoutingTable` with `FailCount == 0` and a fresh `LastSeen` from the last Ping/Verack. Later `FindClosest`/`Sample` will return a dead contact until something else happens to call `MarkFailure`.

Call `MarkFailure` from `PostStop` when `Version` was set (or from `Tcp.ConnectionClosed` before `Context.Stop`).

File: src/Neo/NeoSystem.cs [hunk 5]
```diff
@@ -139,15 +141,16 @@ public NeoSystem(ProtocolSettings settings, string? storageProvider = null, stri
...
+        public NeoSystem(ProtocolSettings settings, IStoreProvider storageProvider, string? storagePath = null, KeyPair? nodeKey = null)
         {
             Settings = settings;
             GenesisBlock = CreateGenesisBlock(settings);
             StorageProvider = storageProvider;
             _store = storageProvider.GetStore(storagePath);
             MemPool = new MemoryPool(this);
             Blockchain = ActorSystem.ActorOf(Ledger.Blockchain.Props(this));
-            LocalNode = ActorSystem.ActorOf(Network.P2P.LocalNode.Props(this));
+            LocalNode = ActorSystem.ActorOf(Network.P2P.LocalNode.Props(this, nodeKey ?? new()));
```
Comment (cschuchardt88) [inline src/Neo/NeoSystem.cs:153]: **suggestion:** `nodeKey ?? new()` mints a random secp256r1 key every process. `NodeId` therefore changes on every restart, which is equivalent to the old per-process `Nonce` for duplicate detection but not for a DHT: other nodes keep stale contacts under the previous id.

Nothing in this repo persists the key. If signed identity is going in, load it from config/wallet; if DHT is not actually used to dial yet, keep the random nonce and do not pretend this is a stable node id.

File: src/Neo/Network/P2P/Payloads/VersionPayload.cs [hunk 6]
```diff
@@ -112,17 +136,34 @@ void ISerializable.Deserialize(ref MemoryReader reader)
             if (capabilities.Select(p => p.Type).Distinct().Count() != capabilities.Count())
                 throw new FormatException("Duplicating capabilities are included");
 
+            Signature = reader.ReadVarMemory().ToArray();
```
Comment (cschuchardt88) [inline src/Neo/Network/P2P/Payloads/VersionPayload.cs:139]: **suggestion:** `ReadVarMemory()` defaults to 16MB. The signature produced by `Crypto.Sign` is 64 bytes, and `Crypto.VerifySignature` throws `FormatException` on any other length (it does not return false).

A Version message can force a 16MB allocate on handshake before verify fails. Use `ReadVarMemory(64)` (or reject `Signature.Length != 64` in `Verify` without throwing through the `bool` API).

File: tests/Neo.UnitTests/Network/P2P/Payloads/UT_VersionPayload.cs [hunk 7]
```diff
@@ -24,40 +23,30 @@ public class UT_VersionPayload
...
+            Assert.AreEqual(148, test.Size);
 
-            test = VersionPayload.Create(123, 456, "neo3", new NodeCapability[] { new ServerCapability(NodeCapabilityType.TcpServer, 22) });
-            Assert.AreEqual(25, test.Size);
+            test = VersionPayload.Create(ProtocolSettings.Default, new(), "neo3", new NodeCapability[] { new ServerCapability(NodeCapabilityType.TcpServer, 22) });
+            Assert.AreEqual(151, test.Size);
         }
 
         [TestMethod]
         public void DeserializeAndSerialize()
```
Comment (cschuchardt88) [inline tests/Neo.UnitTests/Network/P2P/Payloads/UT_VersionPayload.cs:34]: **suggestion:** This no longer covers `Verify`, and it dropped the extra-`UnknownCapability` deserialize case instead of updating offsets for the new layout.

Please add tests for: valid `Create` round-trip `Verify == true`; wrong `NodeId`; signature from another key; empty/short signature (today this throws `FormatException` from `Crypto.VerifySignature` rather than returning false); and the previous unknown-capability tail still deserializing.

**Review by cschuchardt88 — COMMENTED** [review]
## Summary

Backports signed node identity (`NodeKey`/`NodeId`/`Signature` on `VersionPayload`) from #4420 and Kademlia k-bucket primitives from #4422, plus the #4547 closest-node XOR-distance fix. `LocalNode.Nonce` is replaced by a per-process `KeyPair`, and handshake success/failure is fed into `RoutingTable`.

The third commit's `FindClosest` change is correct (full XOR sort before truncation). The rest is not shippable on N3 as-is: the version payload is a hard wire break with `ProtocolVersion` still 0, the signature is replayable, default identity is ephemeral, and the table is never used for peer selection. Maintainer comments asking for a transition period are still unaddressed. PR also conflicts with current `master-n3` on `KeyPair.cs`.

`NeedMorePeers` still uses `GetAddr` / seed list; `RoutingTable.FindClosest` is unused. Breaking the N3 handshake for an unused table matches the existing "Why?" on this PR.

## Issues

- **bug** `src/Neo/Network/P2P/Payloads/VersionPayload.cs:127` — Version handshake is a hard network break (`Nonce` replaced with no dual-decode / version bump). Status: open
- **bug** `src/Neo/Network/P2P/Payloads/VersionPayload.cs:160` — Version signature is replayable; not bound to this connection. Status: open
- **bug** `src/Neo/Network/P2P/RemoteNode.cs:143` — Remote TCP close never marks DHT failure. Status: open
- **suggestion** `src/Neo/NeoSystem.cs:153` — Default node identity is ephemeral (`nodeKey ?? new()`). Status: open
- **suggestion** `src/Neo/Network/P2P/Payloads/VersionPayload.cs:139` — Version signature length is unbounded (`ReadVarMemory()`). Status: open
- **suggestion** `tests/Neo.UnitTests/Network/P2P/Payloads/UT_VersionPayload.cs:34` — No `Verify` coverage; UnknownCapability round-trip dropped. Status: open
